seismicpro/metrics/metric_map.py

```diff
@@ -5,73 +5,161 @@
 from matplotlib import colors as mcolors
 
 from .interactive_map import ScatterMapPlot, BinarizedMapPlot
-from .utils import parse_coords, parse_metric_values
+from .utils import parse_coords, parse_index, parse_metric_data
 from ..decorators import plotter
 from ..utils import to_list, get_first_defined, add_colorbar, calculate_axis_limits, set_ticks, set_text_formatting
 
 
-class BaseMetricMap:
-    """Base metric map class. Implements general input processing, map visualization and reaggregation methods.
+class BaseMetricMap:  # pylint: disable=too-many-instance-attributes
+    """Base metric map class. Implements general map data processing, visualization and reaggregation methods.
 
     Should not be instantiated directly, use `MetricMap` or its subclasses instead.
     """
-    def __init__(self, coords, metric_values, *, coords_cols=None, metric=None, metric_name=None, agg=None):
-        from .metrics import Metric, PartialMetric  # pylint: disable=import-outside-toplevel
-        if metric is None:
-            metric = Metric
-        if not (isinstance(metric, (Metric, PartialMetric)) or
-                isinstance(metric, type) and issubclass(metric, Metric)):
-            raise ValueError("metric must be either of a Metric type or a subclass of Metric")
-
+    def __init__(self, coords, values, *, coords_cols=None, index=None, index_cols=None, metric=None, agg=None,
+                 calculate_immediately=True):
         coords, coords_cols = parse_coords(coords, coords_cols)
-        metric_values, metric_name = parse_metric_values(metric_values, metric_name, metric)
+        values, metric = parse_metric_data(values, metric)
         metric_data = pd.DataFrame(coords, columns=coords_cols)
-        metric_data[metric_name] = metric_values
-
-        self.metric_data = metric_data.dropna()
-        self.coords_cols = coords_cols
-        self.metric_name = metric_name
+        metric_data[metric.name] = values
 
-        if isinstance(metric, Metric):
-            self.metric = metric
-            self.metric.name = metric_name
-            self.metric.coords_cols = coords_cols
+        if index is None:
+            index_cols = coords_cols
         else:
-            self.metric = PartialMetric(metric, name=metric_name, coords_cols=coords_cols)
+            index, index_cols = parse_index(index, index_cols)
+            metric_data[index_cols] = index
 
         if agg is None:
             default_agg = {True: "max", False: "min", None: "mean"}
-            agg = default_agg[self.metric.is_lower_better]
+            agg = default_agg[metric.is_lower_better]
+
+        self._metric_data = None
+        self._index_data = None
+        self._map_data = None
+        self._map_coords_to_indices = None
```

Comment:
To be more consistent with attribute bellow
```suggestion
        self._coords_to_indices = None
```

Reply:
Here I want to emphasize that these coordinates are not the ones passed to a metric constructor: they are the ones displayed on the map plot. E.g. for a binarized map this will map coordinates of bins into its indices.